kolibri/plugins/coach/assets/src/composables/useExerciseResources.js

```diff
@@ -0,0 +1,198 @@
+import { ContentNodeKinds } from 'kolibri.coreVue.vuex.constants';
+import { getContentNodeThumbnail } from 'kolibri.utils.contentNode';
+import { set } from '@vueuse/core';
+// import pickBy from 'lodash/pickBy';
+import uniq from 'lodash/uniq';
+// import store from 'kolibri.coreVue.vuex.store';
+
+export function useExerciseResources() {
+  const resources = ref(null);
+  const channels = ref([]);
+  const bookmarks = ref([]);
+  const channelTopics = ref([]);
+  const contentList = ref([]);
+  const ancestors = ref([]);
+  const currentTopicId = ref(null);
+  const currentTopic = ref(null);
+  const currentTopicResource = ref(null);
+
+  function fetchChannelResource() {
+    ChannelResource.fetchCollection({ params: { has_exercises: true, available: true } }).then(
+      response => {
+        set(
+          channels,
+          response.map(chnl => {
+            return {
+              ...chnl,
+              id: chnl.root,
+              title: chnl.name,
+              kind: ContentNodeKinds.CHANNEL,
+              is_leaf: false,
+            };
+          })
+        );
+      }
+    );
+  }
+
+  function fetchBookMarkedResource() {
+    ContentNodeResource.fetchBookmarks({ params: { limit: 25, available: true } }).then(data => {
+      bookmarks.value = data.results ? data.results : [];
+    });
+  }
+
+  function _getTopicsWithExerciseDescendants(topicIds = []) {
+    return new Promise(resolve => {
+      if (!topicIds.length) {
+        resolve([]);
+        return;
+      }
+      const topicsNumAssessmentDescendantsPromise = ContentNodeResource.fetchDescendantsAssessments(
+        topicIds
+      );
+
+      topicsNumAssessmentDescendantsPromise.then(response => {
+        const topicsWithExerciseDescendants = [];
+        response.data.forEach(descendantAssessments => {
+          if (descendantAssessments.num_assessments > 0) {
+            topicsWithExerciseDescendants.push({
+              id: descendantAssessments.id,
+              numAssessments: descendantAssessments.num_assessments,
+              exercises: [],
+            });
+          }
+        });
+
+        ContentNodeResource.fetchDescendants(
+          topicsWithExerciseDescendants.map(topic => topic.id),
+          {
+            descendant_kind: ContentNodeKinds.EXERCISE,
+          }
+        ).then(response => {
+          response.data.forEach(exercise => {
+            channelTopics.value.push(exercise);
+            const topic = topicsWithExerciseDescendants.find(t => t.id === exercise.ancestor_id);
+            topic.exercises.push(exercise);
+          });
+          channels.value = channelTopics.value;
+          resolve(topicsWithExerciseDescendants);
+        });
+      });
+    });
+  }
+
+  function fetchTopicResource(topicId) {
+    const topicNodePromise = ContentNodeResource.fetchModel({ id: topicId });
+    const childNodesPromise = ContentNodeResource.fetchCollection({
```

Comment:
As per our conversation - this is where we could use the new fetchTree method instead, to fetch both the topic and its children in one go.

To limit data, can pass the `depth: 1` getParam to only get the immediate descendants. The only thing to be aware of here is that the return data here is that the children are paginated by default.

Can see an example of it in use here: https://github.com/learningequality/kolibri/blob/b4233535e112fa3318fbd8bba912b22340614c35/kolibri/plugins/learn/assets/src/views/TopicsPage/index.vue#L426